process/rating/peerHonesty/peerScore.go

```diff
@@ -0,0 +1,280 @@
+	cache storage.Cacher,
+) (*p2pPeerHonesty, error) {
+	err := checkParams(peerHonestyConfig, blackListedPkCache, cache)
+	if err != nil {
+		return nil, fmt.Errorf("%w while creating an instance of p2pPeerHonesty", err)
+	}
+
+	instance := &p2pPeerHonesty{
+		decayCoefficient:       peerHonestyConfig.DecayCoefficient,
+		updateIntervalForDecay: time.Duration(peerHonestyConfig.DecayUpdateIntervalInSeconds) * time.Second,
+		maxScore:               peerHonestyConfig.MaxScore,
+		minScore:               peerHonestyConfig.MinScore,
+		badPeerThreshold:       peerHonestyConfig.BadPeerThreshold,
+		unitValue:              peerHonestyConfig.UnitValue,
+		cache:                  cache,
+		blackListedPkCache:     blackListedPkCache,
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	instance.cancelFunc = cancelFunc
+
+	go instance.executeDecay(ctx, instance.applyDecay)
+
+	return instance, nil
+}
+
+func checkParams(
+	peerHonestyConfig config.PeerHonestyConfig,
+	blackListedPkCache process.TimeCacher,
+	cache storage.Cacher,
+) error {
+	if check.IfNil(blackListedPkCache) {
+		return process.ErrNilBlackListedPkCache
+	}
+
+	if check.IfNil(cache) {
+		return process.ErrNilCacher
+	}
+
+	isDecayCoefficientOk := peerHonestyConfig.DecayCoefficient > minDecayCoefficient &&
+		peerHonestyConfig.DecayCoefficient < maxDecayCoefficient
+	if !isDecayCoefficientOk {
+		return fmt.Errorf("%w, decay coefficient should be in interval (%.2f, %.2f)",
+			process.ErrInvalidDecayCoefficient,
+			minDecayCoefficient,
+			maxDecayCoefficient,
+		)
+	}
+
+	if peerHonestyConfig.DecayUpdateIntervalInSeconds < minDecayIntervalInSeconds {
+		return fmt.Errorf("%w, decay interval in seconds should be greater or equal to %d",
+			process.ErrInvalidDecayIntervalInSeconds,
+			minDecayIntervalInSeconds,
+		)
+	}
+	if peerHonestyConfig.MinScore > 0 {
+		return fmt.Errorf("%w, MinScore value should be negative or zero",
+			process.ErrInvalidMinScore,
+		)
+	}
+	if peerHonestyConfig.MaxScore < 0 {
+		return fmt.Errorf("%w, MaxScore value should be positive or zero",
+			process.ErrInvalidMaxScore,
+		)
+	}
+	if peerHonestyConfig.UnitValue < 0 {
+		return fmt.Errorf("%w, UnitValue value should be positive or zero",
+			process.ErrInvalidUnitValue,
+		)
+	}
+
+	isBadPeerThresholdOk := peerHonestyConfig.BadPeerThreshold < 0 && peerHonestyConfig.MinScore < peerHonestyConfig.BadPeerThreshold
+	if !isBadPeerThresholdOk {
+		return fmt.Errorf("%w, BadPeerThreshold value should be in interval (MinScore, 0)",
+			process.ErrInvalidBadPeerThreshold,
+		)
+	}
+
+	return nil
+}
+
+func (pph *p2pPeerHonesty) executeDecay(ctx context.Context, handler func()) {
+	for {
+		select {
+		case <-time.After(pph.updateIntervalForDecay):
+			handler()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (pph *p2pPeerHonesty) applyDecay() {
+	pph.mut.Lock()
+	defer pph.mut.Unlock()
+
+	keys := pph.cache.Keys()
+	for _, key := range keys {
+		psObj, _ := pph.cache.Get(key)
```

Comment:
why don't you treat the error instead of nil check ?

Reply:
done